sql: Add support for specifying window frames for window functions

yuzefovich opened the pull request:
WIP on #26464:
ROWS mode is fully supported whereas RANGE works only with
UNBOUNDED PRECEDING/CURRENT ROW/UNBOUNDED FOLLOWING boundaries
(same as in PostgreSQL). Current implementation of aggregate functions
is naive (it simply computes the value of aggregate directly and
discards all the previous computations which results in quadratic time).

Release note: None

yuzefovich committed:
sql: Add support for specifying window frames for window functions

WIP on #26464:
ROWS mode is fully supported whereas RANGE works only with
UNBOUNDED PRECEDING/CURRENT ROW/UNBOUNDED FOLLOWING boundaries
(same as in PostgreSQL). Current implementation of aggregate functions
is naive (it simply computes the value of aggregate directly and
discards all the previous computations which results in quadratic time).

Release note (sql change): CockroachDB now supports custom frame
specification for window functions using ROWS (fully-supported)
and RANGE ('value' PRECEDING and 'value' FOLLOWING are not supported)
modes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26666)
<!-- Reviewable:end -->

knz commented:
(nit: As this is a user-visible change you'll need to populate a release note in the commit message(s) before this merges.)

knz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/sem/tree/walk.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuMjOL0Pfp3rLoU6y7:-LEuMjOL0Pfp3rLoU6y8:bo15nc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/walk.go#L217)):*
> ```Go
> 			windowFrameCopy := *windowDef.Frame
> 			windowDef.Frame = &windowFrameCopy
> 			if windowFrameBoundsCopy := *windowDef.Frame.Bounds; &windowFrameBoundsCopy != nil {
> ```

this if is problematic. I think you mean this:
```go
if windowDef.Frame.Bounds != nil {
    windowFrameBoundsCopy := *windowDef.Frame.Bounds
    windowDef.Frame.Bounds = &windowFrameBoundsCopy
}
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 9 of 10 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/window.go, line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuPgZpEKNad_dIhS6a:-LEuPgZpEKNad_dIhS6b:b-rwx2gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L170)):*
> ```Go
> 		bounds := frame.Bounds
> 		if bounds.StartBound.BoundType == tree.ValuePreceding || bounds.StartBound.BoundType == tree.ValueFollowing {
> 			typedStartOffset := bounds.StartBound.TypedOffsetExpr
> ```

Do this instead:
`typedStarOffset := bounds.StartBound.OffsetExpr.(tree.TypedExpr)`

no need for a separate field. Ditto below.

---

*[pkg/sql/window.go, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuOFzeBMDxFiUtYhuo:-LEuOFzeBMDxFiUtYhup:b-slql3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L280)):*
> ```Go
> 				funcHolder.funcIdx = prevWindowCount + i
> 				funcHolder.argIdxStart = len(newRenders)
> 				funcHolder.windowDef = *funcHolder.expr.WindowDef
> ```

You're making a copy of WindowDef here. If you are making this copy before the calls to analyzeExpr() in `constructWindowDefinitions`, then later `funcHolder.windowDef` will not contain the proper expressions. This is the main reason why you encounter the subquery error.

Remember that 

1) expression trees, once constructed, must be immutable. We reuse trees across multiple executions of a statement so the parsed tree must never be mutated in-place.

2) `analyzeExpr` rewrites the expression tree to produce the typed expression. (You can verify this by checking that the address of the root expression node has changed.)

I believe you need to analyze the WindowDef just once (and all its sub-expressions) and then memoize the result somehow.

---

*[pkg/sql/window.go, line 357 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuQMcU9gioA0z5yI-X:-LEuQMcU9gioA0z5yI-Y:b-hdc5g6) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L357)):*
> ```Go
> 		}
> 
> 		n.run.windowFrame = windowDef.Frame
> ```

Make a copy of *both* the frame and bounds object because you are mutating them below. As per my comment above the original syntax tree must remain immutable.

---

*[pkg/sql/parser/sql.y, line 7177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuP_vQ-DiM0dvckEKl:-LEuP_vQ-DiM0dvckEKm:b-hs3c6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/parser/sql.y#L7177)):*
> ```Yacc
>     switch {
>     case startBound.BoundType == tree.UnboundedFollowing:
>       sqllex.Error("frame start cannot be UNBOUNDED FOLLOWING")
> ```

This needs both positive and negative tests in `parse_test.go`.

---

*[pkg/sql/sem/tree/window_funcs.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuOofD1An0tX6k9pm5:-LEuOofD1An0tX6k9pm6:b-4l2wdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L54)):*
> ```Go
> 	Offset          int
> 	OffsetExpr      Expr
> 	TypedOffsetExpr TypedExpr
> ```

Don't make a separate field for the typed expression. Reuse `OffsetExpr` directly.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/window.go, line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuPgZpEKNad_dIhS6a:-LEuRs1M3JvMvYTh1aRH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L170)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Do this instead:
`typedStarOffset := bounds.StartBound.OffsetExpr.(tree.TypedExpr)`

no need for a separate field. Ditto below.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuOFzeBMDxFiUtYhuo:-LEutHww04IfEaQMnLmm:b2ex13x) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L280)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

You're making a copy of WindowDef here. If you are making this copy before the calls to analyzeExpr() in `constructWindowDefinitions`, then later `funcHolder.windowDef` will not contain the proper expressions. This is the main reason why you encounter the subquery error.

Remember that 

1) expression trees, once constructed, must be immutable. We reuse trees across multiple executions of a statement so the parsed tree must never be mutated in-place.

2) `analyzeExpr` rewrites the expression tree to produce the typed expression. (You can verify this by checking that the address of the root expression node has changed.)

I believe you need to analyze the WindowDef just once (and all its sub-expressions) and then memoize the result somehow.
</blockquote></details>

Thanks for the detailed explanation - it helped me find a bug (I called `analyzeExpr` on a copy of parsed tree but later was using the original parsed tree).

---

*[pkg/sql/window.go, line 357 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuQMcU9gioA0z5yI-X:-LEutHLZBvxhqJee4mZG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L357)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make a copy of *both* the frame and bounds object because you are mutating them below. As per my comment above the original syntax tree must remain immutable.
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 7177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuP_vQ-DiM0dvckEKl:-LEv3OOA69DR90JS_tVb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/parser/sql.y#L7177)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This needs both positive and negative tests in `parse_test.go`.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/walk.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuMjOL0Pfp3rLoU6y7:-LEuRVSqDmCQ8VCnpRgE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/walk.go#L217)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

this if is problematic. I think you mean this:
```go
if windowDef.Frame.Bounds != nil {
    windowFrameBoundsCopy := *windowDef.Frame.Bounds
    windowDef.Frame.Bounds = &windowFrameBoundsCopy
}
```
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/window_funcs.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEuOofD1An0tX6k9pm5:-LEuRSg-DtFmv3uQ06L1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L54)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Don't make a separate field for the typed expression. Reuse `OffsetExpr` directly.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 6 of 10 files at r1, 1 of 6 files at r2.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/window.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv57Z43Nz2CYt1QBeN:-LEv57Z43Nz2CYt1QBeO:b-rwrye8) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L165)):*
> ```Go
> 
> 	// OffsetExpr's must be integer expressions not containing any variables, aggregate functions, or window functions,
> 	// so we need to make sure these expressions are evaluated before using offsets.
> ```

Where are you enforcing this? `MustBeDInt` will panic if the resulting datum is not a `DInt`, but I don't see any other type checking that forces the type (just one that desired it).

---

*[pkg/sql/window.go, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv5WpY6LlAJIUHBEvC:-LEv5WpY6LlAJIUHBEvD:b2f04dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L169)):*
> ```Go
> 	if frame != nil && frame.Bounds != nil {
> 		bounds := frame.Bounds
> 		if bounds.StartBound.BoundType == tree.ValuePreceding || bounds.StartBound.BoundType == tree.ValueFollowing {
> ```

This is leaking an abstraction. Either make `HasExpr` a method on `WindowFrameBoundType` or just check if the expression is nil or not.

---

*[pkg/sql/logictest/testdata/logic_test/window, line 1518 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6ADXEVixrj_xCnaH:-LEv6ADXEVixrj_xCnaI:b-ct83p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/logictest/testdata/logic_test/window#L1518)):*
> ```
> SELECT final_variance(1.2, 1.2, 123) OVER (PARTITION BY k) FROM kv
> ```

I'm assuming you tested all of this against PG as well?

---

*[pkg/sql/logictest/testdata/logic_test/window, line 1543 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv64884Czm8YtLhjEL:-LEv64884Czm8YtLhjEM:b1vsofl) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/logictest/testdata/logic_test/window#L1543)):*
> ```
> ('Tablet', 'Samsung', 200, 200, 200)
> 
> statement error cannot copy window "w" because it has a frame clause
> ```

Add tests where you're given a non-int start and end bounds.

---

*[pkg/sql/parser/sql.y, line 7183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6PZL64QHj0zKhFmg:-LEv6PZL64QHj0zKhFmh:b8jznwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/parser/sql.y#L7183)):*
> ```Yacc
>       return 1
>     case startBound.BoundType == tree.CurrentRow && endBound.BoundType == tree.ValuePreceding:
>       sqllex.Error("frame starting from current row cannot have preceding rows")
> ```

Are these errors straight from PG?

---

*[pkg/sql/sem/builtins/window_builtins.go, line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6la10CAvghIIRO95:-LEv6la10CAvghIIRO96:bocss18) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6141bb9b7ba82af63b5c4f62fa43b620a9484b6/pkg/sql/sem/builtins/window_builtins.go#L181)):*
> ```Go
> }
> 
> // framableAggregateWindowFunc is a wrapper around aggregateWindowFunc that allows
> ```

Will this abstraction be useful when we look to break the current quadratic behavior by performing some memoization using a segment tree?

---

*[pkg/sql/sem/tree/window_funcs.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv42mo2SXlkW9grbzS:-LEv42mo2SXlkW9grbzT:b-ax9j8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

For the sake of reviewability, I'd strongly suggest you go back and make the filename fix in a commit where you touch nothing else.

---

*[pkg/sql/sem/tree/window_funcs.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv00l9ALcpsx4U8JhX:-LEv00l9ALcpsx4U8JhY:b-lvmkvq) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L52)):*
> ```Go
> type WindowFrameBound struct {
> 	BoundType       WindowFrameBoundType
> 	Offset          int
> ```

I'm also not sure we need this. OffsetExpr should be all we need.

---

*[pkg/sql/sem/tree/window_funcs.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv0guhAYtqazMdcvha:-LEv0guhAYtqazMdcvhb:bxxujn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L60)):*
> ```Go
> // The row at StartBound is included whereas the row at EndBound is not.
> type WindowFrameBounds struct {
> 	StartBound *WindowFrameBound
> ```

Does this need to be nullable? Looks like it's always non-nil if `WindowFrameBounds` is non-nil.

---

*[pkg/sql/sem/tree/window_funcs.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv3Aud2Y9qLZecRPio:-LEv3Aue-c1Wk6eAcJxU:bahdi90) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L65)):*
> ```Go
> 
> // WindowFrame is a view into a subset of data over which calculations are made.
> type WindowFrame struct {
> ```

I think we're misconstruing two concepts here. `WindowFrame` is and was used as an object to hold state between calls to a `WindowFunc` interface during evaluation. It's not an AST node, which for all intents and purposes should be immutable and contain no dynamic state. We're going to want to lift a lot of the new code here up to live next to `WindowDef`.

---

*[pkg/sql/sem/tree/window_funcs.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv14M014hz304ZmISF:-LEv14M014hz304ZmISG:bipx1pj) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L86)):*
> ```Go
> 		ctx.WriteString("UNBOUNDED PRECEDING")
> 	case ValuePreceding:
> 		ctx.WriteString(fmt.Sprintf("%v PRECEDING", boundary.Offset))
> ```

Replace here and below with
```
ctx.FormatNode(boundary.OffsetExpr)
ctx.WriteString(" PRECEDING")
```

---

*[pkg/sql/sem/tree/window_funcs.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv1dwF3ka0AHW5gKnZ:-LEv1dwF3ka0AHW5gKn_:b9iez86) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L93)):*
> ```Go
> 	case UnboundedFollowing:
> 		ctx.WriteString("UNBOUNDED FOLLOWING")
> 	}
> ```

Go doesn't statically enforce that switch statements are exhaustive. It's generally good practice to do so with a default case like:
```
default:
    panic("unexpected")
```

I'd add those throughout.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/window.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv57Z43Nz2CYt1QBeN:-LEyq4Ya-UqvCV93AYDX:b-a2at29) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L165)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Where are you enforcing this? `MustBeDInt` will panic if the resulting datum is not a `DInt`, but I don't see any other type checking that forces the type (just one that desired it).
</blockquote></details>

`TypeCheckAndRequire` was called as a part of `analyzeExpr` call within `constructWindowDefinitions`. However, with Nathan's help I realized that `analyzeExpr` is not necessary and was there because I had a bug earlier. Hopefully, new version is the way to go.

---

*[pkg/sql/window.go, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv5WpY6LlAJIUHBEvC:-LEyyYwdACnJ5cbsvKTL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/window.go#L169)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is leaking an abstraction. Either make `HasExpr` a method on `WindowFrameBoundType` or just check if the expression is nil or not.
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/window, line 1518 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6ADXEVixrj_xCnaH:-LEykY9PAdp1fL1kUSeI:b-v2exn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/logictest/testdata/logic_test/window#L1518)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm assuming you tested all of this against PG as well?
</blockquote></details>

Yes, the only differences I found in the output are the number of zeros that PG and CRDB return in number like 200.00 and printing out NULL values (PG leaves them empty), but I think that these are expected behavior, right?

---

*[pkg/sql/logictest/testdata/logic_test/window, line 1543 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv64884Czm8YtLhjEL:-LEyzTQ26Ce2SmNuFEJ0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/logictest/testdata/logic_test/window#L1543)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add tests where you're given a non-int start and end bounds.
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 7183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6PZL64QHj0zKhFmg:-LEv8GtU19ngT7M6kxLS:bgkwxse) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/parser/sql.y#L7183)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are these errors straight from PG?
</blockquote></details>

Yes, I manually tried all these combination to get their error messages so that ours would be consistent with them.

---

*[pkg/sql/sem/builtins/window_builtins.go, line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6la10CAvghIIRO95:-LEyzUXrAgWuonFGJ_CE:b-roepbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6141bb9b7ba82af63b5c4f62fa43b620a9484b6/pkg/sql/sem/builtins/window_builtins.go#L181)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Will this abstraction be useful when we look to break the current quadratic behavior by performing some memoization using a segment tree?
</blockquote></details>

I'm not sure, to be honest. I plan on adding a method that would replace current aggregator with faster implementation (via `switch` on type of the current aggregator like `avg`). In that case, `aggConstructor` would be set to `nil` to use advantage of memoization. However, this abstraction allows us to add fast implementations of some (say, `avg`, `min`, `max`) aggregate functions and to go back to slow quadratic behavior if faster is not present.

---

*[pkg/sql/sem/tree/window_funcs.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv42mo2SXlkW9grbzS:-LEyliXXDlAdqfgacDjW:b-ovwaw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L1)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

For the sake of reviewability, I'd strongly suggest you go back and make the filename fix in a commit where you touch nothing else.
</blockquote></details>

Good idea.

---

*[pkg/sql/sem/tree/window_funcs.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv00l9ALcpsx4U8JhX:-LEzNMrs6ol0X-HGpZBF:b-frazj) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm also not sure we need this. OffsetExpr should be all we need.
</blockquote></details>

As we discussed, `OffsetExpr` is evaluated only once and the result is cached in `Offset`.

---

*[pkg/sql/sem/tree/window_funcs.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv0guhAYtqazMdcvha:-LEzNugF2u9JzNg-rarV:be424ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L60)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this need to be nullable? Looks like it's always non-nil if `WindowFrameBounds` is non-nil.
</blockquote></details>

Correct, `StartBound` is always non-nil, but making it an instance instead of a pointer would require changes in the parser (I believe we would need to handle `startBound` and `endBound` differently since they would be returning different types which would require some copy/pasting), so I think it's better to keep it as a pointer. (Possibly I'm not seeing a better way to go about changes in the parser.)

---

*[pkg/sql/sem/tree/window_funcs.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv3Aud2Y9qLZecRPio:-LEzckqD9wcZl0WP05n0:b-mwdkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L65)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think we're misconstruing two concepts here. `WindowFrame` is and was used as an object to hold state between calls to a `WindowFunc` interface during evaluation. It's not an AST node, which for all intents and purposes should be immutable and contain no dynamic state. We're going to want to lift a lot of the new code here up to live next to `WindowDef`.
</blockquote></details>

Ok, I divided it into two structs. Please let me know if further refinements are desirable (and give suggestions if so).

---

*[pkg/sql/sem/tree/window_funcs.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv14M014hz304ZmISF:-LEv9WyD-bZVtE-e7xd5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L86)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Replace here and below with
```
ctx.FormatNode(boundary.OffsetExpr)
ctx.WriteString(" PRECEDING")
```
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/window_funcs.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv1dwF3ka0AHW5gKnZ:-LEvAM5F4-2olfD-MdDI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L93)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Go doesn't statically enforce that switch statements are exhaustive. It's generally good practice to do so with a default case like:
```
default:
    panic("unexpected")
```

I'd add those throughout.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Generally looks good, and nice tests. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once comments get resolved. Looks like you'll need to rebase this against master as there's some merge conflicts.

---

Reviewed 2 of 10 files at r1, 1 of 6 files at r2, 8 of 9 files at r3.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/window.go, line 428 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LF4PWXE7o4Tt3856NWG:-LF4PWXE7o4Tt3856NWH:bwgrwmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/91476b0830bd42906ca4ab87d3ed1816bf600724/pkg/sql/window.go#L428)):*
> ```Go
> 
> 	if referencedSpec.Frame != nil {
> 		return def, errors.Errorf("cannot copy window %q because it has a frame clause", refName)
> ```

you should use a pgerror here, to match Postgres. In `psql`, if you enter `\set VERBOSITY verbose` before running commands, you will see specific error codes depending on the error type. This one happens to return `42P20`, windowing error, which can be retrieved in Cockroach by saying 

`pgerror.NewErrorf(pgerror.CodeWindowingError, "cannot copy window %q because it has a frame clause", refName)`

---

*[pkg/sql/sem/builtins/window_builtins.go, line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6la10CAvghIIRO95:-LF4OPCOAcZqMHwUfTLw:b-vfugcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6141bb9b7ba82af63b5c4f62fa43b620a9484b6/pkg/sql/sem/builtins/window_builtins.go#L181)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I'm not sure, to be honest. I plan on adding a method that would replace current aggregator with faster implementation (via `switch` on type of the current aggregator like `avg`). In that case, `aggConstructor` would be set to `nil` to use advantage of memoization. However, this abstraction allows us to add fast implementations of some (say, `avg`, `min`, `max`) aggregate functions and to go back to slow quadratic behavior if faster is not present.
</blockquote></details>

I think it'll be useful to keep around as well for testing purposes. THis is a straightforward implementation that will be useful as a testing baseline, in case it's difficult to implement the more efficient algorithms.

---

*[pkg/sql/sem/tree/window_funs.go, line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LF4MoYY7qslhXGQo6TY:-LF4MoYY7qslhXGQo6TZ:b-336pga) ([raw file](https://github.com/cockroachdb/cockroach/blob/91476b0830bd42906ca4ab87d3ed1816bf600724/pkg/sql/sem/tree/window_funs.go#L113)):*
> ```Go
> 
> // Copy returns a deep copy of wf.
> func (wf WindowFrame) Copy() *WindowFrame {
> ```

minor: It seems odd that the receiver is not a pointer but it is for Format. I don't know if it matters much, but for consistency I'd make the receiver a pointer as well.

---

*[pkg/sql/sem/tree/window_funs.go, line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LF4N0Bb8xViNl32yWio:-LF4N0Bb8xViNl32yWip:blkb3m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/91476b0830bd42906ca4ab87d3ed1816bf600724/pkg/sql/sem/tree/window_funs.go#L132)):*
> ```Go
> 	ArgIdxStart int // the index which arguments to the window function begin
> 	ArgCount    int // the number of window function arguments
> 	Frame       *WindowFrame
> ```

Add a comment above this new field explaining what this is. If it can be nil, say so - something like `If non-nil, Frame represents the frame specification for this window.`

---

*[pkg/sql/sem/tree/window_funs.go, line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LF4Rcvu4gFL1gD5CA6e:-LF4Rcvu4gFL1gD5CA6f:b-h2819o) ([raw file](https://github.com/cockroachdb/cockroach/blob/91476b0830bd42906ca4ab87d3ed1816bf600724/pkg/sql/sem/tree/window_funs.go#L153)):*
> ```Go
> 			return 0
> 		case ValuePreceding:
> 			// TODO(yuzefovich): Currently, it is not supported, and this case should not be reached.
> ```

minor: I would just panic in these situations instead of returning `0`, to make it blindingly clear that these won't be reached, but it's not a big deal either way.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-:-LF4RzHo6thJDYMTSiN6:b-3z0t49)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/window.go, line 428 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LF4PWXE7o4Tt3856NWG:-LF4bPd-CjCVhmdyFude:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91476b0830bd42906ca4ab87d3ed1816bf600724/pkg/sql/window.go#L428)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

you should use a pgerror here, to match Postgres. In `psql`, if you enter `\set VERBOSITY verbose` before running commands, you will see specific error codes depending on the error type. This one happens to return `42P20`, windowing error, which can be retrieved in Cockroach by saying 

`pgerror.NewErrorf(pgerror.CodeWindowingError, "cannot copy window %q because it has a frame clause", refName)`
</blockquote></details>

Done.

---

*[pkg/sql/sem/builtins/window_builtins.go, line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6la10CAvghIIRO95:-LF4cEdpEH7kG9vKrGNl:btd51a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6141bb9b7ba82af63b5c4f62fa43b620a9484b6/pkg/sql/sem/builtins/window_builtins.go#L181)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think it'll be useful to keep around as well for testing purposes. THis is a straightforward implementation that will be useful as a testing baseline, in case it's difficult to implement the more efficient algorithms.
</blockquote></details>

I agree.

---

*[pkg/sql/sem/tree/window_funs.go, line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LF4MoYY7qslhXGQo6TY:-LF4ZQnOC3lwbT8fZLe7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91476b0830bd42906ca4ab87d3ed1816bf600724/pkg/sql/sem/tree/window_funs.go#L113)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

minor: It seems odd that the receiver is not a pointer but it is for Format. I don't know if it matters much, but for consistency I'd make the receiver a pointer as well.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/window_funs.go, line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LF4N0Bb8xViNl32yWio:-LF4ZF5P3gir3yWDMTit:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91476b0830bd42906ca4ab87d3ed1816bf600724/pkg/sql/sem/tree/window_funs.go#L132)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Add a comment above this new field explaining what this is. If it can be nil, say so - something like `If non-nil, Frame represents the frame specification for this window.`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
This is looking better and better. Nice job so far.

---

Reviewed 10 of 11 files at r4.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/window.go, line 357 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJgcRs0LHh4NgN3NC1:-LFJgcRs0LHh4NgN3NC2:bf61mmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/window.go#L357)):*
> ```Go
> 		n.run.windowFrame = windowDef.Frame
> 		// Validate window frame bounds if present
> 		frame := n.run.windowFrame
> ```

Why are we type checking again? We shouldn't need to perform any type checking in this file.

---

*[pkg/sql/logictest/testdata/logic_test/window, line 1518 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv6ADXEVixrj_xCnaH:-LFJgII95vqh2ubMjaHx:b-rnt45) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/logictest/testdata/logic_test/window#L1518)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yes, the only differences I found in the output are the number of zeros that PG and CRDB return in number like 200.00 and printing out NULL values (PG leaves them empty), but I think that these are expected behavior, right?
</blockquote></details>

Yep, both of those are expected divergences.

---

*[pkg/sql/sem/builtins/window_builtins.go, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJfwuu0znwqa2MDJhN:-LFJfwuu0znwqa2MDJhO:b-jx5r4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/builtins/window_builtins.go#L206)):*
> ```Go
> 	// and construct a new one for the computation.
> 	w.agg.Close(ctx, evalCtx)
> 	w.agg = &aggregateWindowFunc{agg: w.aggConstructor(evalCtx)}
> ```

Don't reallocate this, just reassign the value: `*w.agg = aggregateWindowFunc{...}`

---

*[pkg/sql/sem/builtins/window_builtins.go, line 234 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJhllA5Nm0HnUfs3kd:-LFJhllA5Nm0HnUfs3ke:b-yugqu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/builtins/window_builtins.go#L234)):*
> ```Go
> }
> 
> // AddAggregateConstructorToFramableAggregate add provided constructor to framableAggregateWindowFunc
> ```

Is this modifying the builtin itself? Is that safe? Aren't these static objects?

---

*[pkg/sql/sem/builtins/window_builtins.go, line 239 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJfd0BEhgR-enyKgh8:-LFJfd0BEhgR-enyKgh9:b-gi26zf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/builtins/window_builtins.go#L239)):*
> ```Go
> 	windowFunc tree.WindowFunc, aggConstructor func(*tree.EvalContext) tree.AggregateFunc,
> ) {
> 	if framableAgg, ok := windowFunc.(*framableAggregateWindowFunc); ok {
> ```

It's ok if `windowFunc` is not a `*framableAggregateWindowFunc`?

---

*[pkg/sql/sem/tree/type_check.go, line 765 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJeoZP8dPqtKPtR4yx:-LFJeoZP8dPqtKPtR4yy:beoy7he) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/tree/type_check.go#L765)):*
> ```Go
> 			startBound, endBound := bounds.StartBound, bounds.EndBound
> 			if startBound.OffsetExpr != nil {
> 				typedStartOffsetExpr, err := startBound.OffsetExpr.TypeCheck(ctx, types.Int)
> ```

I'm still not sure how we're getting away with not replacing this line with
```
typedStartOffsetExpr, err := typeCheckAndRequire(ctx, startBound.OffsetExpr, types.Int, "window frame offset")
```

---

*[pkg/sql/sem/tree/window_funs.go, line 268 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJeco78ctDJmSbuRsM:-LFJeco78ctDJmSbuRsN:bmsbocq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/tree/window_funs.go#L268)):*
> ```Go
> }
> 
> func (wfr WindowFrameRun) unboundedFollowing() int {
> ```

Add a comment to this method to explain it's meaning.

---

*[pkg/sql/sem/tree/window_funcs.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv42mo2SXlkW9grbzS:-LFJcEHa2XZ2VrADpirj:b-7sdli7) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L1)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Good idea.
</blockquote></details>

Thanks, that helps a lot.

---

*[pkg/sql/sem/tree/window_funcs.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv00l9ALcpsx4U8JhX:-LFJcpQE4bPhtefOVfQl:banknkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L52)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

As we discussed, `OffsetExpr` is evaluated only once and the result is cached in `Offset`.
</blockquote></details>

But then `Offset` should be part `WindowFrameRun`.

---

*[pkg/sql/sem/tree/window_funcs.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv0guhAYtqazMdcvha:-LFJdxD42bxo6MOw44jC:b-1tf30h) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L60)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Correct, `StartBound` is always non-nil, but making it an instance instead of a pointer would require changes in the parser (I believe we would need to handle `startBound` and `endBound` differently since they would be returning different types which would require some copy/pasting), so I think it's better to keep it as a pointer. (Possibly I'm not seeing a better way to go about changes in the parser.)
</blockquote></details>

Ok, that makes sense.

---

*[pkg/sql/sem/tree/window_funcs.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv3Aud2Y9qLZecRPio:-LFJcP4f4DnZ2IVwA81l:bq5l4ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L65)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Ok, I divided it into two structs. Please let me know if further refinements are desirable (and give suggestions if so).
</blockquote></details>

Thanks! I would suggest that you collocate `WindowFrame` and all of its related AST types with `Window` and `WindowDef`. I don't have a strong opinion about which file they're in, only that they're in the same file.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-:-LFJhzuE4-od9YABmrp5:b1juxru)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/window.go, line 357 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJgcRs0LHh4NgN3NC1:-LFJnKOJ-ZrIrMK70MaT:blfehon) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/window.go#L357)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why are we type checking again? We shouldn't need to perform any type checking in this file.
</blockquote></details>

You're right - calling `TypeCheckAndRequire` within `type_check.go` should be sufficient.

---

*[pkg/sql/sem/builtins/window_builtins.go, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJfwuu0znwqa2MDJhN:-LFJoDx9DINw4X8iadC9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/builtins/window_builtins.go#L206)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Don't reallocate this, just reassign the value: `*w.agg = aggregateWindowFunc{...}`
</blockquote></details>

Done.

---

*[pkg/sql/sem/builtins/window_builtins.go, line 234 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJhllA5Nm0HnUfs3kd:-LFJoMkTEm13XaYMgm2F:bupx6tm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/builtins/window_builtins.go#L234)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this modifying the builtin itself? Is that safe? Aren't these static objects?
</blockquote></details>

The builtins are not modified, and I believe this to be safe - we simply create a new instance of a builtin (if necessary) to achieve "resetting" behavior. As far as I understand, these objects are not static, i.e. we can create as many as we want without new ones influencing the old ones.

---

*[pkg/sql/sem/builtins/window_builtins.go, line 239 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJfd0BEhgR-enyKgh8:-LFJrLb3-PFdtKbYl7zI:b-ch7svg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/builtins/window_builtins.go#L239)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's ok if `windowFunc` is not a `*framableAggregateWindowFunc`?
</blockquote></details>

Yes, it's ok since I've amended all builtins specific to window functions (i.e. non-aggregates, but functions like `rank`) to support window frames "natively". In a sense, those functions are not framable aggregates, so they should not be of that type. I added this comment to the code.

---

*[pkg/sql/sem/tree/type_check.go, line 765 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJeoZP8dPqtKPtR4yx:-LFJsSD31dxOs2Nwmo2z:b7qo7vy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/tree/type_check.go#L765)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm still not sure how we're getting away with not replacing this line with
```
typedStartOffsetExpr, err := typeCheckAndRequire(ctx, startBound.OffsetExpr, types.Int, "window frame offset")
```
</blockquote></details>

Thanks for pointing it out. Indeed, type checking and requiring `Int` in here is sufficient.

---

*[pkg/sql/sem/tree/window_funs.go, line 268 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFJeco78ctDJmSbuRsM:-LFJtX1c7pMW-EKZtNIa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a04a1a260b0482a6db073845de365fad0c6914/pkg/sql/sem/tree/window_funs.go#L268)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a comment to this method to explain it's meaning.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/window_funcs.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv42mo2SXlkW9grbzS:-LFJk9Ep0-twbUx8eaYJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L1)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Thanks, that helps a lot.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/window_funcs.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv00l9ALcpsx4U8JhX:-LFJvNL9CuwolCWpg5Za:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

But then `Offset` should be part `WindowFrameRun`.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/window_funcs.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LEv3Aud2Y9qLZecRPio:-LFJwnAG2RAMUp_ihdkw:bb97c2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/939b7f50c252d8cc6fad2596f204e1e0fc2c04fb/pkg/sql/sem/tree/window_funcs.go#L65)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Thanks! I would suggest that you collocate `WindowFrame` and all of its related AST types with `Window` and `WindowDef`. I don't have a strong opinion about which file they're in, only that they're in the same file.
</blockquote></details>

I moved everything into `select.go` since window functions can only be used with `SELECT`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 7 files at r5.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

---

*[pkg/sql/sem/tree/type_check.go, line 765 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFrXQeZ3pgjLkCQ1Haj:-LFrXQeZ3pgjLkCQ1Hak:blnc6w6) ([raw file](https://github.com/cockroachdb/cockroach/blob/433d30cc56ce63bd76a623c923a03edee89defb8/pkg/sql/sem/tree/type_check.go#L765)):*
> ```Go
> 			startBound, endBound := bounds.StartBound, bounds.EndBound
> 			if startBound.OffsetExpr != nil {
> 				typedStartOffsetExpr, err := TypeCheckAndRequire(startBound.OffsetExpr, ctx, types.Int, "window frame start")
> ```

I'm not sure when `TypeCheckAndRequire` (exported) was introduced, but this should use `typeCheckAndRequire` (not exported).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-:-LFrXujT04UpF2KBOboB:bnfp4nl)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

---

*[pkg/sql/sem/tree/type_check.go, line 765 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-LFrXQeZ3pgjLkCQ1Haj:-LFrf3L-3DAU8J2Y8hBK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/433d30cc56ce63bd76a623c923a03edee89defb8/pkg/sql/sem/tree/type_check.go#L765)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm not sure when `TypeCheckAndRequire` (exported) was introduced, but this should use `typeCheckAndRequire` (not exported).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26666#-:-LFrf5Gt5MWswO9ujZrC:bz5maj4)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+

yuzefovich commented:
bors r+
